Storage/OAuthStorage.php

```diff
@@ -164,12 +170,22 @@ public function checkUserCredentials(IOAuth2Client $client, $username, $password
         }
 
         if (null !== $user) {
-            $encoder = $this->encoderFactory->getEncoder($user);
-
-            if ($encoder->isPasswordValid($user->getPassword(), $password, $user->getSalt())) {
-                return array(
-                    'data' => $user,
-                );
+            $data = array(
+                'data' => $user,
+            );
+
+            if ($this->tokenPathAuth === 'plain') {
+                $encoder = $this->encoderFactory->getEncoder($user);
+
+                if ($encoder->isPasswordValid($user->getPassword(), $password, $user->getSalt())) {
+                    return $data;
+                }
+            } elseif ($this->tokenPathAuth === 'hash') {
+                if ($user->getPassword() === $password) {
```

Comment:
Using this comparison, the client is supposed to know the salted password stored in the storage.
This is not really secured: if the pasword of the user is `SJ%-qgjC]o/NWtID[u:HI#7,#Yy\d97~(DiNEPLKgyL;x?2|S` and the salted password in the storage is `8e604e3823eeec0e3209f36e7a1e5f9eeacd46b7797d1db60cc74f82aae65901` (sha256 of `SaltSJ%-qgjC]o/NWtID[u:HI#7,#Yy\d97~(DiNEPLKgyL;x?2|S`), the credential used by the client is less secured than the password.